File: schema/analysis.go
// Package schema defines the canonical data types for the repo-mri analysis output.
package schema

import "time"

// SchemaVersion is the current version of the analysis.json schema.
const SchemaVersion = "1.1"

// CLIVersion is the current version of the CLI.
const CLIVersion = "0.1.0"

// Analysis is the top-level output artifact written to .repo-mri/analysis.json.
type Analysis struct {
	Meta         Meta         `json:"meta"`
	Repo         Repo         `json:"repo"`
	Modules      []Module     `json:"modules"`
	Dependencies []Dependency `json:"dependencies"`
	Risks        []Risk       `json:"risks"`
	Files        []File       `json:"files"`
	Subsystems   []Subsystem  `json:"subsystems,omitempty"`
}

// Meta holds metadata about the analysis run.
type Meta struct {
	SchemaVersion      string   `json:"schema_version"`
	CLIVersion         string   `json:"cli_version"`
	ModelUsed          string   `json:"model_used,omitempty"`
	Provider           string   `json:"provider,omitempty"`
	AnalysisDurationMS int64    `json:"analysis_duration_ms"`
	SkippedPasses      []string `json:"skipped_passes,omitempty"`
	// MaxChainDepth is the length of the longest module dependency chain. Set by Phase 2.
	MaxChainDepth int `json:"max_chain_depth,omitempty"`
}

// Repo holds repository-level summary data.
type Repo struct {
	Name         string    `json:"name"`
	Languages    []string  `json:"languages"`
	FileCount    int       `json:"file_count"`
	ModuleCount  int       `json:"module_count"`
	AnalysisTime time.Time `json:"analysis_time"`
}

// Module represents a logical grouping of files (typically a package/directory).
type Module struct {
	ID              string  `json:"id"`
	Path            string  `json:"path"`
	Language        string  `json:"language"`
	FileCount       int     `json:"file_count"`
	RiskScore       float64 `json:"risk_score"`
	ComplexityScore float64 `json:"complexity_score"`
	// ImportCount is the number of other modules that import this module. Set by Phase 2.
	ImportCount int `json:"import_count,omitempty"`
}

// Dependency represents a directed import relationship between modules.
type Dependency struct {
	From string `json:"from"`
	To   string `json:"to"`
	Type string `json:"type"`
}

// Risk represents a single finding from an analysis pass.
type Risk struct {
	ID            string  `json:"id"`
	Severity      string  `json:"severity"`
	Type          string  `json:"type"`
	Pass          string  `json:"pass"`
	Module        string  `json:"module"`
	File          string  `json:"file"`
	Title         string  `json:"title"`
	Description   string  `json:"description"`
	Confidence    float64 `json:"confidence"`
	EvidenceLines []int   `json:"evidence_lines,omitempty"`
}

// File holds per-file metrics.
type File struct {
	Path     string `json:"path"`
	Module   string `json:"module"`
	Language string `json:"language"`
	Lines    int    `json:"lines"`
	// Size is the file size in bytes. Set by Phase 2.
	Size       int64   `json:"size,omitempty"`
	Complexity float64 `json:"complexity"`
	RiskScore  float64 `json:"risk_score"`
}

// Subsystem groups related modules.
type Subsystem struct {
	ID      string   `json:"id"`
	Modules []string `json:"modules"`
}

File: ui/src/types/analysis.ts
/**
 * TypeScript mirror of schema/analysis.go (schema version 1.2).
 * Keep in sync when the CLI schema version changes.
 */

/** Top-level output artifact written to .repo-mri/analysis.json. */
export interface Analysis {
  meta: Meta
  repo: Repo
  modules: Module[]
  dependencies: Dependency[]
  risks: Risk[]
  files: File[]
  subsystems?: Subsystem[]
}

/** Metadata about the analysis run. */
export interface Meta {
  schema_version: string
  cli_version: string
  model_used?: string
  provider?: string
  analysis_duration_ms: number
  skipped_passes?: string[]
  /** Length of the longest module dependency chain. Set by Phase 2. */
  max_chain_depth?: number
  /**
   * Absolute filesystem path to the repository root.
   * Set for local analyses only; absent for remote (GitHub) analyses.
   * Required to construct working VS Code deep links.
   */
  root_path?: string
}

/** Repository-level summary data. */
export interface Repo {
  name: string
  /** Full "owner/repo" slug, populated only when analyzed from a GitHub URL. */
  github_slug?: string
  languages: string[]
  file_count: number
  module_count: number
  analysis_time: string
}

/** Logical grouping of files (typically a package/directory). */
export interface Module {
  id: string
  path: string
  language: string
  file_count: number
  risk_score: number
  complexity_score: number
  /** Number of other modules that import this module. Set by Phase 2. */
  import_count?: number
}

/** Directed import relationship between modules. */
export interface Dependency {
  from: string
  to: string
  type: string
}

/** Single finding from an analysis pass. */
export interface Risk {
  id: string
  severity: string
  type: string
  pass: string
  module: string
  file: string
  title: string
  description: string
  confidence: number
  evidence_lines?: number[]
  /** Classifies what the finding targets: "file", "module", or "repository". */
  target_type?: string
  /** File path, module ID, or repo name depending on target_type. */
  target_id?: string
}

/** Per-file metrics. */
export interface File {
  path: string
  module: string
  language: string
  lines: number
  /** File size in bytes. Set by Phase 2. */
  size?: number
  complexity: number
  risk_score: number
}

/** Groups related modules. */
export interface Subsystem {
  id: string
  modules: string[]
}
